src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs

```diff
@@ -306,6 +308,18 @@ private IEnumerable<PackageSource> LoadNuGetSources(PackageSourceLocation packag
             PackageSourceProvider packageSourceProvider = new PackageSourceProvider(settings);
             defaultSources = packageSourceProvider.LoadPackageSources().Where(source => source.IsEnabled);
 
+            PackageSourceMapping packageSourceMapping = PackageSourceMapping.GetPackageSourceMapping(settings);
+
+            // filter package patterns if enabled            
+            if (packageSourceMapping?.IsEnabled == true)
+            {
+                IReadOnlyList<string> sources = packageSourceMapping.GetConfiguredPackageSources(packageId.ToString());
+
+                if (sources.Count == 0)
+                {
+                    throw new NuGetPackageInstallerException("No NuGet sources are defined or enabled");
```

Comment:
This error message should be a localized string.  The message should probably be something like:

> Package Source Mapping is enabled but no feeds matched the specified packed ID: {0}

@nkolev92 @zivkan @baronfel Thoughts on what this message should say?

We should also add tests for this failure case, as well as some basic other scenarios when package source mapping is enabled.

Reply:
The NuGet.Client repo has package source mapping enabled, and when I try `dotnet tool install -g zivkan.doesnotexist`, I get this output, with the current restore-based tool install:

> C:\Users\zivkan\AppData\Local\Temp\d0755727-b84b-4758-b239-72d92b6c2d9c\restore.csproj : error NU1100: Unable to resolv
e 'zivkan.doesnotexist (>= 0.0.0)' for 'net8.0'. PackageSourceMapping is enabled, the following source(s) were not cons
idered: dotnet-eng, dotnet-public, myget-legacy@Local, nuget-build, vside.
> C:\Users\zivkan\AppData\Local\Temp\d0755727-b84b-4758-b239-72d92b6c2d9c\restore.csproj : error NU1100: Unable to resolv
e 'zivkan.doesnotexist (>= 0.0.0)' for 'net8.0/any'. PackageSourceMapping is enabled, the following source(s) were not
considered: dotnet-eng, dotnet-public, myget-legacy@Local, nuget-build, vside.
> The tool package could not be restored.
> Tool 'zivkan.doesnotexist' failed to install. This failure may have been caused by:
> 
> * You are attempting to install a preview release and did not use the --version option to specify the version.
> * A package by this name was found, but it was not a .NET tool.
> * The required NuGet feed cannot be accessed, perhaps because of an Internet connection problem.
> * You mistyped the name of the tool.
> 
> For more reasons, including package naming enforcement, visit https://aka.ms/failure-installing-tool

I don't know if you want to keep the NU1100 error code. [Our docs for NU1100](https://learn.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu1100#scenario-2) already tell customers to check their package source mapping config in nuget.config. But if you don't use more or less the exact same message, then maybe it'll be confusing to customers who ctrl-f on the page to find snippets of their error message, without actually reading the docs.

Anyway, my point is that NuGet adds " PackageSourceMapping is enabled, the following source(s) were not cons
idered: dotnet-eng, dotnet-public, myget-legacy@Local, nuget-build, vside." to the end of the existing message. So, whatever message you're planning to use for `dotnet tool install -g ToolDoesNotExist`, you can append this message to the end when package source mapping is enabled.

Reply:
I don't think that we should attempt to re-use the NuGet.Client message directly unless it is a publicly accessible resource that we could include via code. I just checked and it's an `internal static` so I wouldn't recommend that route. Instead, we should explain the problem and guide the user to the existing package source mapping documentation in this case:

> Package Source Mapping is enabled for this project but no package sources were configured for the package '{0}'.  See the documentation for Package Source Mapping at https://learn.microsoft.com/nuget/consume-packages/package-source-mapping#enable-by-manually-editing-nugetconfig for more details.

What context will this string be shown in? Will it be part of a larger message? 

### Separate topic, Chet going off the rails a bit now 

I think we can be much more authoritative with our guidance for users when the install fails now that we're using the NuGet APIs directly. Let's take the 4 options we currently show during package installation failures that @zivkan showed us above:

> You are attempting to install a preview release and did not use the --version option to specify the version.

We should be able to detect this now - if we're querying for versions of a particular package we can check if all of the versions are prereleases and if so tell the user that this is 100% the problem.

> A package by this name was found, but it was not a .NET tool.

Similarly, we can now detect this and say 100% that this is the problem

> The required NuGet feed cannot be accessed, perhaps because of an Internet connection problem.

We should also get exceptions directly from NuGet that are connectivity related and so could be more direct.

> You mistyped the name of the tool.

this is the only one that I think we couldn't detect - we could try a fallback package name search and return the top N from the configured package sources though in an attempt to be more helpful?